File: ts/src/MbtaV3SDK.ts
// MbtaV3 Ts SDK

import { AlertEntity } from './entity/AlertEntity'
import { FacilityEntity } from './entity/FacilityEntity'
import { LineEntity } from './entity/LineEntity'
import { PredictionEntity } from './entity/PredictionEntity'
import { RouteEntity } from './entity/RouteEntity'
import { RoutePatternEntity } from './entity/RoutePatternEntity'
import { ScheduleEntity } from './entity/ScheduleEntity'
import { ServiceEntity } from './entity/ServiceEntity'
import { ShapeEntity } from './entity/ShapeEntity'
import { StopEntity } from './entity/StopEntity'
import { TripEntity } from './entity/TripEntity'
import { VehicleEntity } from './entity/VehicleEntity'

export type * from './MbtaV3Types'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { MbtaV3EntityBase } from './MbtaV3EntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class MbtaV3SDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _alert?: AlertEntity

  // Idiomatic facade: `client.alert.list()` / `client.alert.load({ id })`.
  get alert(): AlertEntity {
    return (this._alert ??= new AlertEntity(this, undefined))
  }

  /** @deprecated Use `client.alert` instead. */
  Alert(data?: any) {
    const self = this
    return new AlertEntity(self,data)
  }


  _facility?: FacilityEntity

  // Idiomatic facade: `client.facility.list()` / `client.facility.load({ id })`.
  get facility(): FacilityEntity {
    return (this._facility ??= new FacilityEntity(this, undefined))
  }

  /** @deprecated Use `client.facility` instead. */
  Facility(data?: any) {
    const self = this
    return new FacilityEntity(self,data)
  }


  _line?: LineEntity

  // Idiomatic facade: `client.line.list()` / `client.line.load({ id })`.
  get line(): LineEntity {
    return (this._line ??= new LineEntity(this, undefined))
  }

  /** @deprecated Use `client.line` instead. */
  Line(data?: any) {
    const self = this
    return new LineEntity(self,data)
  }


  _prediction?: PredictionEntity

  // Idiomatic facade: `client.prediction.list()` / `client.prediction.load({ id })`.
  get prediction(): PredictionEntity {
    return (this._prediction ??= new PredictionEntity(this, undefined))
  }

  /** @deprecated Use `client.prediction` instead. */
  Prediction(data?: any) {
    const self = this
    return new PredictionEntity(self,data)
  }


  _route?: RouteEntity

  // Idiomatic facade: `client.route.list()` / `client.route.load({ id })`.
  get route(): RouteEntity {
    return (this._route ??= new RouteEntity(this, undefined))
  }

  /** @deprecated Use `client.route` instead. */
  Route(data?: any) {
    const self = this
    return new RouteEntity(self,data)
  }


  _route_pattern?: RoutePatternEntity

  // Idiomatic facade: `client.route_pattern.list()` / `client.route_pattern.load({ id })`.
  get route_pattern(): RoutePatternEntity {
    return (this._route_pattern ??= new RoutePatternEntity(this, undefined))
  }

  /** @deprecated Use `client.route_pattern` instead. */
  RoutePattern(data?: any) {
    const self = this
    return new RoutePatternEntity(self,data)
  }


  _schedule?: ScheduleEntity

  // Idiomatic facade: `client.schedule.list()` / `client.schedule.load({ id })`.
  get schedule(): ScheduleEntity {
    return (this._schedule ??= new ScheduleEntity(this, undefined))
  }

  /** @deprecated Use `client.schedule` instead. */
  Schedule(data?: any) {
    const self = this
    return new ScheduleEntity(self,data)
  }


  _service?: ServiceEntity

  // Idiomatic facade: `client.service.list()` / `client.service.load({ id })`.
  get service(): ServiceEntity {
    return (this._service ??= new ServiceEntity(this, undefined))
  }

  /** @deprecated Use `client.service` instead. */
  Service(data?: any) {
    const self = this
    return new ServiceEntity(self,data)
  }


  _shape?: ShapeEntity

  // Idiomatic facade: `client.shape.list()` / `client.shape.load({ id })`.
  get shape(): ShapeEntity {
    return (this._shape ??= new ShapeEntity(this, undefined))
  }

  /** @deprecated Use `client.shape` instead. */
  Shape(data?: any) {
    const self = this
    return new ShapeEntity(self,data)
  }


  _stop?: StopEntity

  // Idiomatic facade: `client.stop.list()` / `client.stop.load({ id })`.
  get stop(): StopEntity {
    return (this._stop ??= new StopEntity(this, undefined))
  }

  /** @deprecated Use `client.stop` instead. */
  Stop(data?: any) {
    const self = this
    return new StopEntity(self,data)
  }


  _trip?: TripEntity

  // Idiomatic facade: `client.trip.list()` / `client.trip.load({ id })`.
  get trip(): TripEntity {
    return (this._trip ??= new TripEntity(this, undefined))
  }

  /** @deprecated Use `client.trip` instead. */
  Trip(data?: any) {
    const self = this
    return new TripEntity(self,data)
  }


  _vehicle?: VehicleEntity

  // Idiomatic facade: `client.vehicle.list()` / `client.vehicle.load({ id })`.
  get vehicle(): VehicleEntity {
    return (this._vehicle ??= new VehicleEntity(this, undefined))
  }

  /** @deprecated Use `client.vehicle` instead. */
  Vehicle(data?: any) {
    const self = this
    return new VehicleEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new MbtaV3SDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return MbtaV3SDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'MbtaV3' }
  }

  toString() {
    return 'MbtaV3 ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = MbtaV3SDK


export {
  stdutil,

  BaseFeature,
  MbtaV3EntityBase,

  MbtaV3SDK,
  SDK,
}

File: go/core/mbta-v3_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/mbta-v3-sdk/go/utility/struct"
)

type MbtaV3SDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewMbtaV3SDK(options map[string]any) *MbtaV3SDK {
	sdk := &MbtaV3SDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features in the resolved order (MakeOptions puts an explicit array
	// order first, else defaults to test-first). Ordering matters: the `test`
	// feature installs the base mock transport and the transport features
	// (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
	// must be added before them to sit at the base of the chain.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		if fo, ok := vs.GetPath([]any{"__derived__", "featureorder"}, sdk.options).([]any); ok {
			for _, n := range fo {
				fname, _ := n.(string)
				fopts := ToMapAny(featureOpts[fname])
				if fopts != nil {
					if active, ok := fopts["active"]; ok {
						if ab, ok := active.(bool); ok && ab {
							sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
						}
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *MbtaV3SDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *MbtaV3SDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *MbtaV3SDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *MbtaV3SDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *MbtaV3SDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Alert returns a Alert entity bound to this client.
// Idiomatic usage: client.Alert(nil).List(nil, nil) or
// client.Alert(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Alert(data map[string]any) MbtaV3Entity {
	return NewAlertEntityFunc(sdk, data)
}


// Facility returns a Facility entity bound to this client.
// Idiomatic usage: client.Facility(nil).List(nil, nil) or
// client.Facility(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Facility(data map[string]any) MbtaV3Entity {
	return NewFacilityEntityFunc(sdk, data)
}


// Line returns a Line entity bound to this client.
// Idiomatic usage: client.Line(nil).List(nil, nil) or
// client.Line(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Line(data map[string]any) MbtaV3Entity {
	return NewLineEntityFunc(sdk, data)
}


// Prediction returns a Prediction entity bound to this client.
// Idiomatic usage: client.Prediction(nil).List(nil, nil) or
// client.Prediction(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Prediction(data map[string]any) MbtaV3Entity {
	return NewPredictionEntityFunc(sdk, data)
}


// Route returns a Route entity bound to this client.
// Idiomatic usage: client.Route(nil).List(nil, nil) or
// client.Route(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Route(data map[string]any) MbtaV3Entity {
	return NewRouteEntityFunc(sdk, data)
}


// RoutePattern returns a RoutePattern entity bound to this client.
// Idiomatic usage: client.RoutePattern(nil).List(nil, nil) or
// client.RoutePattern(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) RoutePattern(data map[string]any) MbtaV3Entity {
	return NewRoutePatternEntityFunc(sdk, data)
}


// Schedule returns a Schedule entity bound to this client.
// Idiomatic usage: client.Schedule(nil).List(nil, nil) or
// client.Schedule(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Schedule(data map[string]any) MbtaV3Entity {
	return NewScheduleEntityFunc(sdk, data)
}


// Service returns a Service entity bound to this client.
// Idiomatic usage: client.Service(nil).List(nil, nil) or
// client.Service(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Service(data map[string]any) MbtaV3Entity {
	return NewServiceEntityFunc(sdk, data)
}


// Shape returns a Shape entity bound to this client.
// Idiomatic usage: client.Shape(nil).List(nil, nil) or
// client.Shape(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Shape(data map[string]any) MbtaV3Entity {
	return NewShapeEntityFunc(sdk, data)
}


// Stop returns a Stop entity bound to this client.
// Idiomatic usage: client.Stop(nil).List(nil, nil) or
// client.Stop(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Stop(data map[string]any) MbtaV3Entity {
	return NewStopEntityFunc(sdk, data)
}


// Trip returns a Trip entity bound to this client.
// Idiomatic usage: client.Trip(nil).List(nil, nil) or
// client.Trip(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Trip(data map[string]any) MbtaV3Entity {
	return NewTripEntityFunc(sdk, data)
}


// Vehicle returns a Vehicle entity bound to this client.
// Idiomatic usage: client.Vehicle(nil).List(nil, nil) or
// client.Vehicle(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *MbtaV3SDK) Vehicle(data map[string]any) MbtaV3Entity {
	return NewVehicleEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *MbtaV3SDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewMbtaV3SDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}
